File: tennis/tennis_20210228/code.py
class TennisGame(object):
    score_map = {
        0: 'Love',
        1: 'Fifteen',
        2: 'Thirty',
        3: 'Forty',
    }

    def __init__(self, player_one_name, player_two_name) -> None:
        super().__init__()

        self.player_one_name = player_one_name
        self.player_two_name = player_two_name
        self.player_one_score = None
        self.player_two_score = None

    def set_player_one_score(self, score):
        self.player_one_score = score

    def set_player_two_score(self, score):
        self.player_two_score = score

    def score(self):
        if self.player_one_score == self.player_two_score:
            score = self.player_one_score

            if score >= 3:
                return 'Deuce'
            else:
                return f'{self.score_map[score]} All'

        if self.player_one_score >= 5:
            return f'{self.player_one_name} Win'

        if self.player_two_score >= 5:
            return f'{self.player_two_name} Win'

        if self.player_one_score >= 4:
            return f'{self.player_one_name} Adv'

        if self.player_two_score >= 4:
            return f'{self.player_two_name} Adv'

        return ' '.join(
            [
                self.score_map[self.player_one_score],
                self.score_map[self.player_two_score],
            ]
        )

File: tennis/tennis_20210228/code.ts
export default class TennisGame {
  private _playerTwoScore = 0;
  private _playerOneScore = 0;

  private map = new Map<number, string>();

  constructor(private playerOneName: string, private playerTwoName: string) {
    this.map.set(0, 'Love');
    this.map.set(1, 'Fifteen');
    this.map.set(2, 'Thirty');
    this.map.set(3, 'Forty');
  }

  score() {
    if (this.isTie()) {
      if (this.isDeuce()) {
        return 'Deuce';
      } else {
        return `${this.playerOneScoreName} All`;
      }
    } else {
      if (this._playerOneScore === 5) {
        return `${this.playerOneName} Win`;
      }
      if (this._playerTwoScore === 5) {
        return `${this.playerTwoName} Win`;
      }
      if (this._playerOneScore === 4) {
        return `${this.playerOneName} Adv`;
      }
      if (this._playerTwoScore === 4) {
        return `${this.playerTwoName} Adv`;
      }

      return `${this.playerOneScoreName} ${this.playerTwoScoreName}`;
    }
  }

  private isDeuce() {
    return this._playerOneScore >= 3;
  }

  private isTie() {
    return this._playerOneScore === this._playerTwoScore;
  }

  get playerOneScoreName() {
    return this.map.get(this._playerOneScore);
  }

  get playerTwoScoreName() {
    return this.map.get(this._playerTwoScore);
  }

  playerOneScore() {
    this._playerOneScore += 1;
  }

  playerTwoScore() {
    this._playerTwoScore += 1;
  }
}
